presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFilePageReader.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.pagefile;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.page.PagesSerde;
+import com.facebook.presto.spi.page.SerializedPage;
+import io.airlift.slice.InputStreamSliceInput;
+import io.airlift.slice.SliceInput;
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.facebook.presto.spi.page.PagesSerdeUtil.readSerializedPage;
+import static com.facebook.presto.spi.page.PagesSerdeUtil.serializedPageWrittenBytes;
+import static java.util.Objects.requireNonNull;
+
+public class PageFilePageReader
+        implements Iterator<Page>
+{
+    private final PagesSerde pagesSerde;
+    private final SliceInput input;
+    private final long readLength;
+    private long readBytes;
+
+    PageFilePageReader(
+            long readStart,
+            long readLength,
+            FSDataInputStream inputStream,
+            PagesSerde pagesSerde)
+            throws IOException
+    {
+        this.pagesSerde = requireNonNull(pagesSerde, "pagesSerde is null");
+        this.readLength = readLength;
+        requireNonNull(inputStream, "inputStream is null");
+        inputStream.seek(readStart);
+        this.input = new InputStreamSliceInput(inputStream);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        return readBytes < readLength && input.isReadable();
```

Comment:
Instead of keeping track of `readBytes` you can use `ByteStreams#limit`

Reply:
Thanks you,  I get rid of `readBytes` using `InputStreamSliceInput.position()`